include/nbl/builtin/glsl/ext/FFT/fft.glsl

```diff
@@ -0,0 +1,247 @@
+}
+
+uvec3 nbl_glsl_ext_FFT_getPaddedDimensions() {
+    nbl_glsl_ext_FFT_Parameters_t params = nbl_glsl_ext_FFT_getParameters();
+    return (params.padded_dimension.xyz);
+}
+uvec3 nbl_glsl_ext_FFT_getDimensions() {
+    nbl_glsl_ext_FFT_Parameters_t params = nbl_glsl_ext_FFT_getParameters();
+    return (params.dimension.xyz);
+}
+uint nbl_glsl_ext_FFT_getDirection() {
+    nbl_glsl_ext_FFT_Parameters_t params = nbl_glsl_ext_FFT_getParameters();
+    return (params.dimension.w >> 16) & 0x000000ff;
+}
+bool nbl_glsl_ext_FFT_getIsInverse() {
+    nbl_glsl_ext_FFT_Parameters_t params = nbl_glsl_ext_FFT_getParameters();
+    return bool((params.dimension.w >> 8) & 0x000000ff);
+}
+uint nbl_glsl_ext_FFT_getPaddingType() {
+    nbl_glsl_ext_FFT_Parameters_t params = nbl_glsl_ext_FFT_getParameters();
+    return (params.dimension.w) & 0x000000ff;
+}
+
+uint nbl_glsl_ext_FFT_getChannel()
+{
+    uint direction = nbl_glsl_ext_FFT_getDirection();
+    return gl_WorkGroupID[direction];
+}
+
+uvec3 nbl_glsl_ext_FFT_getCoordinates(in uint tidx)
+{
+    uint direction = nbl_glsl_ext_FFT_getDirection();
+    uvec3 tmp = gl_WorkGroupID;
+    tmp[direction] = tidx;
+    return tmp;
+}
+
+uvec3 nbl_glsl_ext_FFT_getBitReversedCoordinates(in uvec3 coords, in uint leadingZeroes)
+{
+    uint direction = nbl_glsl_ext_FFT_getDirection();
+    uint bitReversedIndex = bitfieldReverse(coords[direction]) >> leadingZeroes;
+    uvec3 tmp = coords;
+    tmp[direction] = bitReversedIndex;
+    return tmp;
+}
+
+uint nbl_glsl_ext_FFT_getDimLength(uvec3 dimension)
+{
+    uint direction = nbl_glsl_ext_FFT_getDirection();
+    return dimension[direction];
+}
+
+void nbl_glsl_ext_FFT(bool is_inverse)
+{
+    nbl_glsl_ext_FFT_Parameters_t params = nbl_glsl_ext_FFT_getParameters();
+    // Virtual Threads Calculation
+    uint dataLength = nbl_glsl_ext_FFT_getDimLength(nbl_glsl_ext_FFT_getPaddedDimensions());
+    uint num_virtual_threads = (dataLength-1u)/(_NBL_GLSL_EXT_FFT_WORKGROUP_SIZE_)+1u;
+    uint thread_offset = gl_LocalInvocationIndex;
+
+	uint channel = nbl_glsl_ext_FFT_getChannel();
+    
+	// Pass 0: Bit Reversal
+	uint leadingZeroes = nbl_glsl_clz(dataLength) + 1u;
+	uint logTwo = 32u - leadingZeroes;
+	
+    nbl_glsl_complex current_values[_NBL_GLSL_EXT_FFT_MAX_ITEMS_PER_THREAD];
+    nbl_glsl_complex shuffled_values[_NBL_GLSL_EXT_FFT_MAX_ITEMS_PER_THREAD];
+
+    // Load Initial Values into Local Mem (bit reversed indices)
+    for(uint t = 0u; t < num_virtual_threads; t++)
+    {
+        uint tid = thread_offset + t * _NBL_GLSL_EXT_FFT_WORKGROUP_SIZE_;
+        uvec3 coords = nbl_glsl_ext_FFT_getCoordinates(tid);
+        uvec3 bitReversedCoords = nbl_glsl_ext_FFT_getBitReversedCoordinates(coords, leadingZeroes);
+
+        current_values[t] = nbl_glsl_ext_FFT_getPaddedData(bitReversedCoords, channel);
+    }
+
+    // For loop for each stage of the FFT (each virtual thread computes 1 buttefly wing)
+	for(uint i = 0u; i < logTwo; ++i) 
```

Comment:
my idea for reducing shared memory usage.

Make a `nbl_glsl_ext_workgroup_FFT` function which is specialized in performing an FFT for a `2*_NBL_GLSL_WORKGROUP_SIZE_` sized input (you can use `4*_NBL_GLSL_WORKGROUP_SIZE_ ` shared memory to not have to use so many barriers for the data exchanges)

Then have this function combine `0x1u<<max(findMSB(dataLength-1)-_NBL_GLSL_WORKGROUP_SIZE_LOG_2_,0)` of these "double workgroup sized" FFTs together

Note that in this case your shuffle mask is `1u<j` where `j>_NBL_GLSL_WORKGROUP_SIZE_LOG_2_` so you end up requesting a shuffled value from a shared memory index assigned to the same "physical thread"

Meaning that instead of going through shared memory and worrying about barriers, you can simply do `shuffled_values[t] = current_values[t^(mask>>(_NBL_GLSL_WORKGROUP_SIZE_LOG_2_+1u))];` without any barriers